lib/MatrixGitterBridge.js

```diff
@@ -0,0 +1,302 @@
+"use strict";
+
+var Promise = require("bluebird");
+
+var Gitter = require('node-gitter');
+var GitterRealtimeClient = require('gitter-realtime-client');
+
+var GitterUser = require("./GitterUser");
+
+var RemoteUser = require("matrix-appservice-bridge").RemoteUser;
+var Bridge = require("matrix-appservice-bridge").Bridge;
+
+var MatrixRoom = require("matrix-appservice-bridge").MatrixRoom;
+
+// TODO: maybe we'll extend it later
+var GitterRoom = require("matrix-appservice-bridge").RemoteRoom;
+
+var BridgedRoom = require("./BridgedRoom");
+
+function MatrixGitterBridge(config) {
+  var self = this;
+
+  this._bridgedRoomsByMatrixId = {};
+
+  this._gitter = new Gitter(config.gitter_api_key);
+  this._gitterRealtime = new GitterRealtimeClient.RealtimeClient({token: config.gitter_api_key});
+
+  var bridge = new Bridge({
+    homeserverUrl: config.matrix_homeserver,
+    domain: config.matrix_user_domain,
+    registration: "gitter-registration.yaml",
+    controller: {
+      onUserQuery: function(queriedUser) {
+        return {}; // auto-provision users with no additonal data
+      },
+
+      onEvent: function(req, context) {
+        var event = req.getData();
+
+        if (event.type == "m.room.member" &&
+            event.state_key == bridge.getBot().getUserId()) {
+          var membership = event.content.membership;
+          if (membership == "invite") {
+            self.onBotInvited(event.room_id);
+          }
+          else if (membership == "join") {
+            self.onBotJoined(event.room_id);
+          }
+          else if (membership == "leave") {
+            // TODO: "leave" events might mean we left the room, or got kicked
+            //   while still in invite state - i.e. invite was cancelled
+            //   before we joined.
+            self.onBotLeft(event.room_id);
+          }
+          else {
+            console.log('matrix member event for myself to state ' + membership + 
+                        ' from state ', event);
+          }
+          return;
+        }
+
+        if (event.type !== "m.room.message" || !event.content) {
+          return;
+        }
+
+        if (event.sender == bridge.getBot().getUserId()) {
+          return;
+        }
+
+        console.log('matrix->' + event.room_id + ' from ' + event.user_id + ':', event.content.body)
+
+        var handled = false;
+
+        if (config.matrix_admin_room && event.room_id == config.matrix_admin_room) {
+          self.onMatrixAdminMessage(event);
+          handled = true;
+        }
+
+        var bridgedRoom = self._bridgedRoomsByMatrixId[event.room_id];
+        if (bridgedRoom) {
+          bridgedRoom.onMatrixMessage(event);
+          handled = true;
+        }
+
+        if (!handled) {
+          console.log("  Wasn't expecting this room; ignore");
+        }
+      }
+    }
+  });
+
+  this._bridge = bridge;
+  this._gitterUserId = null;
+
+  this._gitterusers = {};
```

Comment:
Indentation has gone wonky. This needs to be indented (the function looks to continue to `:109`).